### ISSUE TITLE
Event based metrics and prediction output

### ISSUE THREAD

**Issue by leanderme** [sharathadavanne/multichannel-sed-crnn#5]
Hi, thank you for sharing this. 

I have a custom dataset with monophonic events (no overlapping events). The annotation labels are stored in the same format as in this challenge:

`[filename (string)][tab][event onset time in seconds (float)][tab][event offset time in seconds (float)][tab][event label (string)]`

Another difference is, that there is annotation information for the full length of the audio data, i.e. for every frame there is an annotation. Also, the audio files have been zero-padded to the longest one in the dataset and for these areas, a label "silence" was added:

```python
 if len(audio_data) < max_len_samples:
   zero_pad = np.zeros(max_len_samples - len(audio_data))
   audio_data = np.hstack((audio_data, zero_pad))
```
**My question**
I'm trying to adapt the code to use event based metrics for training epochs to save the best model weights using keras callbacks (like implemented [here](https://github.com/DCASE-REPO/dcase2018_baseline/blob/master/task4/task4_crnn.py#L355)). I have tried *a lot*, without success so far.

Also, I'm trying to output the model predictions file by file, like in the format (like [here](https://github.com/DCASE-REPO/dcase2018_baseline/blob/master/task4/task4_crnn.py#L771))
``` python
{
  'filename': filename,
  'event_label': label,
  'onset': onset * hop_length,
  'offset': offset * hop_length
}
```
but I have no clue how to do that yet. What I tried:

```python
preds_all = model.predict(X_test)
for j, pred in enumerate(preds_all):
  top_pred_indexes = np.argmax(pred,axis=1)
  for z, top_index in enumerate(top_pred_indexes): # len = Frame sequence length
    pred_probability = pred[z][top_index]
    print(j, z, list(__class_labels.keys())[top_index], pred_probability)
```
In summary, I just need to apply event based metrics per epoch to save the weights and find a way to test the model output to find out how the model performs.

Any help is greatly appreciated.

**Comment by sharathadavanne** [comment]
Hi, thanks for checking the repository. 

With regard to using event-based metric - the implementation is straight forward.  As shown in the [link you suggested](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/task4_crnn.py#L669) you pass your trained model to the evaluation metric functions defined [here in the same repository](https://github.com/DCASE-REPO/dcase2018_baseline/blob/master/task4/evaluation_measures.py). Within these functions, since you are not using a data generator, you dont really need the[ for loop](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/evaluation_measures.py#L42). You can straight away compute the statistics (TP, FP, .. ) and calculate the score.

 Regarding the output format, if you are using the DCASE_util repository similar to the repository you referred, then you have to make sure you use all the important components from [line 771](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/task4_crnn.py#L771) to [line 822](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/task4_crnn.py#L822). Else just write a custom code to dump your data into a text file, which might be simpler for you.

**Comment by leanderme** [comment]
Thank you for the fast reply!
I know this has little to do with the repository itself, and I really appreciate your help.

I tried what you suggested, but I have to admit that I don't understand what I'm doing here. I got rid of the for loop, passed in X_Test and Y_Test to calculate the f measure for validation, and the output is always 0. To be explicit, here is the code:

```python
val_macro_f_measure = get_f_measure_by_class(
  model,
  len(list(set(__class_labels.keys()))),
  X_test, Y_test
)
val_macro_f_measure = val_macro_f_measure.mean()

def get_f_measure_by_class(keras_model, nb_tags, X, y, thresholds=None):
  # Calculate external metrics
  TP = numpy.zeros(nb_tags)
  TN = numpy.zeros(nb_tags)
  FP = numpy.zeros(nb_tags)
  FN = numpy.zeros(nb_tags)
      
  predictions = keras_model.predict(X)

  if len(predictions.shape) == 3:
    # average data to have weak labels
    predictions = numpy.mean(predictions, axis=1)
    y = numpy.mean(y, axis=1)

  if thresholds is None:
    binarization_type = 'global_threshold'
    thresh = 0.5
  else:
    binarization_type = "class_threshold"
    assert type(thresholds) is list
    thresh = thresholds

  predictions = ProbabilityEncoder().binarization(
    predictions,
    binarization_type=binarization_type,
    threshold=thresh,
    time_axis=0
  )

  TP += (predictions + y == 2).sum(axis=0)
  FP += (predictions - y == 1).sum(axis=0)
  FN += (y - predictions == 1).sum(axis=0)
  TN += (predictions + y == 0).sum(axis=0)

  macro_f_measure = numpy.zeros(nb_tags)

  mask_f_score = 2 * TP + FP + FN != 0
  macro_f_measure[mask_f_score] = 2*TP[mask_f_score] / (2*TP + FP + FN)[mask_f_score]

  return macro_f_measure
```

The model does not seem to learn anything useful yet:
![model_metrics](https://user-images.githubusercontent.com/4583896/56922652-023c7c80-6ac9-11e9-9bae-a88ebdb14819.png)

But I think this is not the issue at the moment for event based metrics calculation. 

In case you are interested in debugging this, I could publish a repo with the data and the code. It contains data for heart sound segmentation (labels are first and second heart sounds, and systole and diastole). The dataset contains ~ 3000 audio files, with heterogeneous quality and audio length from 5 - 120 seconds.

**Comment by sharathadavanne** [comment]
Hi @leanderme I myself havent worked with event-based metric before. In case you dont know already the SED event and segment based metrics were proposed in [this paper.](https://tutcris.tut.fi/portal/files/6973737/applsci_06_00162.pdf) Check figure 5 and Section 4 for more details on event-base metrics. The segment-based metrics implemented in this repository is shown in figure 4. Personally, I have always used the segment-based metrics, and also most people use only the segment-based metrics for SED task. In case you still want to continue using event-based metrics you can check the official implementation [here.](http://tut-arg.github.io/sed_eval/sound_event.html#module-sed_eval.sound_event)

With respect to the equations in the paper, you can maybe relook into your function and check if the function is doing the right operations. I am attempting a pseudo-code below which is untested. The `predictions` and `y` at the point of calculating the statistics (TP, FP.. ) should be of dimension (TxC). Where T is the number of frames and C is the number of classes. Both these matrices should be in one-hot format.

```python
import utils
def get_f_measure_by_class(keras_model, nb_tags, X, y, thresholds=None):
  # Calculate external metrics
  TP = numpy.zeros(nb_tags)
  TN = numpy.zeros(nb_tags)
  FP = numpy.zeros(nb_tags)
  FN = numpy.zeros(nb_tags)
      
  predictions = keras_model.predict(X)
  predictions = predictions > 0.5

  if len(predictions.shape) == 3:
      predictions, y = utils.reshape_3Dto2D(predictions), utils.reshape_3Dto2D(y)

  TP += (predictions + y == 2).sum(axis=0)
  FP += (predictions - y == 1).sum(axis=0)
  FN += (y - predictions == 1).sum(axis=0)
  TN += (predictions + y == 0).sum(axis=0)

  macro_f_measure = numpy.zeros(nb_tags)

  mask_f_score = 2 * TP + FP + FN != 0
  macro_f_measure[mask_f_score] = 2*TP[mask_f_score] / (2*TP + FP + FN)[mask_f_score]

  return macro_f_measure
```

**Comment by leanderme** [comment]
Thank you for the detailed response! I can confirm that the function works correctly now! I'm entirely new to SED, and I'm greatful for any information. I'll follow your advice and will keep using segment-based metrics during training. 

If you don't mind, could you have a quick look on how to test the model on a random audio file? I've checked the code discussed above [(line 771](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/task4_crnn.py#L771) to [line 822](https://github.com/DCASE-REPO/dcase2018_baseline/blob/f254a638039ebd60f4ad5672a1bf5277c90b61ed/task4/task4_crnn.py#L822), using components of the [dcase_util](https://github.com/DCASE-REPO/dcase_util/)), but I believe this was written to get the onsets and offsets by using a model trained on weak labels (which are tags without time information as labels):

```python
frame_decisions = ProbabilityEncoder().binarization(
  probabilities=preds_all[0],
  binarization_type="global_threshold",
  threshold=0.5,
  time_axis=0
)

decision_encoder = dcase_util.data.DecisionEncoder(
  label_list=list(set(__class_labels.keys()))
)

for k, label in enumerate(list(set(__class_labels.keys()))):
  # given a list of ones, give the onset and offset in frames
  estimated_events = decision_encoder.find_contiguous_regions(
    activity_array=frame_decisions[:, k]
  )
  for [onset, offset] in estimated_events:
    hop_length_seconds = 0.02 #hardcoded for now
    print(
      {
          'filename': audio_filename,
          'event_label': label,
          'onset': onset * hop_length_seconds,
          'offset': offset * hop_length_seconds
      }
    )
```
What I did was:
1. Loaded the audio file (using the same function from the feature extraction stage)
2. Extracted the audio features
3. Applied normalization
4. Applied `split_in_seqs` and `split_multi_channels`
5. Loaded the model definition and the weights

```python
X = mbe
X = utils.split_in_seqs(X, seq_len)
X = utils.split_multi_channels(X, nb_ch)

#Loads model from Json file
model = model_from_json(open("./models/model.json").read())
#Loads pre-trained weights for the model
model.load_weights(weights)
#Compiles the model
model.compile(optimizer='Adam', loss='binary_crossentropy')
model.summary()

```

At the prediction stage, I'm basically iterating the prediction result:

```python
for j, pred in enumerate(preds_all):
  top_pred_indexes = np.argmax(pred,axis=1)
  for z, top_index in enumerate(top_pred_indexes):
    pred_probability = pred[z][top_index]
    if pred_probability >= posterior_thresh:
      print(j,list(__class_labels.keys())[top_index], pred_probability)
```

This way, I get the predictions for 1 splitted segment only and I still don't now how to calculcate the event onset and offset properly. Could you give me a hint on what I'm doing wrong here?

**Comment by sharathadavanne** [comment]
1. Collect all your segment-wise results for individual recordings.
2. If your CRNN model is well trained, then the network should already be able to give you well-separated sound event instances. If not, you will have to use some kind of post-processing to filter out spurious activations occurring in 1-2 frames, maybe by employing a median filter. I would say, dump the CRNN output directly and see how the results look if there are too many spurious activations then use some post-processing.
3. Use the following code to identify the onset-offset times from your file-wise system output matrix obtained in 1. The following code dumps results from all test files into a single CSV file. The `contiguous_regions` function is well tested, but the main script towards the end hasn't been tested. But I guess it should work.


```python
def contiguous_regions(activity_array):
    """Find contiguous regions from bool valued np.array.
    Transforms boolean values for each frame into pairs of onsets and offsets.
    Parameters
    ----------
    activity_array : np.array [shape=(t)]
        Event activity array, bool values
    Returns
    -------
    change_indices : np.ndarray [shape=(2, number of found changes)]
        Onset and offset indices pairs in the matrix
    """

    # Find the changes in the activity_array
    change_indices = np.diff(activity_array).nonzero()[0]

    # Shift change_index with one, focus on the frame after the change.
    change_indices += 1

    if activity_array[0]:
        # If the first element of activity_array is True add 0 at the beginning
        change_indices = np.r_[0, change_indices]

    if activity_array[-1]:
        # If the last element of activity_array is True, add the length of the array
        change_indices = np.r_[change_indices, activity_array.size]

    # Reshape the result into two columns
    return change_indices.reshape((-1, 2))

# MAIN SCRIPT STARTS HERE
hop_length = 0.02
filename = 'models/{}_results.csv'.format(feat_name)
fid = open(filename, 'w')
for filecnt, filename in enumerate(test_file_name_list):
    pred_file = model.predict(test_feat_list[filecnt])
    pred_file_threshold = (pred_file > 0.5).astype(int)
    pred_file_threshold = reshape_3Dto2D(pred_file_threshold)
    for col in range(pred_file_threshold.shape[1]):
	# For each class, identify all onset-offset times of its instances
        region_list = contiguous_regions(pred_file_threshold[:, col])
        for region in region_list:
            fid.write('{}\t{}\t{}\t{}\n'.format(filename, region[0]*hop_length, region[1]*hop_length, class_labels[col]))
fid.close()
```

**Comment by leanderme** [comment]
I'm failing at point 1. I think this is implemented in your [dcase2019 repository](https://github.com/sharathadavanne/seld-dcase2019/) in the [DataGenerator class](https://github.com/sharathadavanne/seld-dcase2019/blob/master/cls_data_generator.py) (using the per_file property) and in the [generate method at line 103](https://github.com/sharathadavanne/seld-dcase2019/blob/master/cls_data_generator.py#L103). Could you give me a hint how to generate the file-wise matrix as input for the model prediction? 

I think the following example code only generates the input matrix for one sequence (`X`), right? I'm struggling finding a way to add all those sequences for one input file together.

```python
y, sr = load_audio(audio_file, mono=is_mono, fs=sr)
# shape (nb_mel_bands, frames)
X = melspectrogram(y, sr, nfft, nb_mel_bands, _hop_len)
# shape (0, seq_len, frames)
X = utils.split_in_seqs(X, seq_len)
# shape (0, nb_channels, seq_len, frames)
X = utils.split_multi_channels(X, nb_ch)
```

Using the generate method as example, I came up with this:
```python
def get_frame_per_file(seq_len, batch_size):
  return batch_size * seq_len

y, sr, silence_entry, original_len = load_audio(audio_file, mono=is_mono, fs=sr)
# shape (nb_mel_bands, feat_len)
temp_feat = melspectrogram(y, sr, nfft, nb_mel_bands, _hop_len)
batch_seq_len = get_frame_per_file(seq_len, batch_size)

circ_buf_feat = deque()
fid = open("results.txt", 'w')
while len(circ_buf_feat) < batch_seq_len:
  feat_len = temp_feat.shape[1]

  for row_cnt, row in enumerate(temp_feat):
    circ_buf_feat.append(row)

  extra_frames = batch_seq_len - temp_feat.shape[0]
  extra_feat = np.ones((extra_frames, temp_feat.shape[1])) * 1e-6

  for row_cnt, row in enumerate(extra_feat):
    circ_buf_feat.append(row)
feat = np.zeros((batch_seq_len, feat_len))


for j in range(batch_seq_len):
  feat[j, :] = circ_buf_feat.popleft()[0]
  feat = np.reshape(feat, (batch_seq_len, feat_len))

feat = utils.split_in_seqs(feat, seq_len)
feat = utils.split_multi_channels(feat, 1)

hop_length = 0.02
pred_file = model.predict(feat)
pred_file_threshold = (pred_file > thresh).astype(int)
pred_file_threshold = utils.reshape_3Dto2D(pred_file_threshold)

# col are the class indices, i.e. 0 1 2 3 4
for col in range(pred_file_threshold.shape[1]):
  # For each class, identify all onset-offset times of its instances
  region_list = contiguous_regions(pred_file_threshold[:, col])
  # for each class there is an array of onset and offset times, f.i. [[ 2 20]]
  for region in region_list:
    fid.write('{}\t{}\t{}\t{}\n'.format(
      filename,
      region[0] * hop_length,
      region[1] * hop_length,
      list(__class_labels.keys())[col])
    )
```
which seems to correctly generate the file-wise input matrix now (f.i. feat has a shape of (64, 1, 20, 12001 now instead of (0, 1, 20, 12001)! But I don't fully understand what the generate method from above does. For example, what is the use of this:

```python
  extra_frames = batch_seq_len - temp_feat.shape[0]
  extra_feat = np.ones((extra_frames, temp_feat.shape[1])) * 1e-6

  for row_cnt, row in enumerate(extra_feat):
    circ_buf_feat.append(row)
  feat = np.zeros((batch_seq_len, feat_len))
```

Also, the results are still strange. For different example files, it predicts one class for almost the entire time. This might be related to the trained model, but could be also caused an error generating the input matrix?

**Comment by sharathadavanne** [comment]
I think you are complicating things. The easier way to do it for this particular repository and dataset would be to first dump the feature normalization weights during feature extraction. Copy paste the following code [here.](https://github.com/sharathadavanne/sed-crnn/blob/da81004c759e5471a22ed287cb95d7105dd7b9f6/feature.py#L204)

```python
from sklearn.externals import joblib
normalized_features_wts_file = '/path/to/wts/filename'

# Save the feature scaler parameters
joblib.dump(
    scaler,
    normalized_features_wts_file
)
```

Now create a new script, that performs testing as following. This is just pseudocode and you might need to add more input variables to run it.
```python

is_mono = True
nb_ch = 1
hop_length = 0.02

filename = 'models/{}_results.csv'.format(feat_name)
fid = open(filename, 'w')
for filecnt, audio_filename in enumerate(test_audiofile_list):
	
	# Load audio
	audio_file = os.path.join(audio_folder, audio_filename)
	y, sr = load_audio(audio_file, mono=is_mono, fs=sr)
	 
	# Extract features
	mbe = extract_mbe(y, sr, nfft, nb_mel_bands).T

	# Load feature scaler
	spec_scaler = joblib.load(normalized_features_wts_file)

	# Scale features
	feat = spec_scaler.transform(mbe)

	# split into sequencse and multichannels
	feat = split_in_seqs(feat, seq_len)
	feat = split_multi_channels(feat, nb_ch)

	# Predict
        pred_file = model.predict(test_feat_list[filecnt])
        pred_file_threshold = (pred_file > 0.5).astype(int)
        pred_file_threshold = reshape_3Dto2D(pred_file_threshold)

        for col in range(pred_file_threshold.shape[1]):
	    # For each class, identify all onset-offset times of its instances
            region_list = contiguous_regions(pred_file_threshold[:, col])
            for region in region_list:
                fid.write('{}\t{}\t{}\t{}\n'.format(audio_filename, region[0]*hop_length, region[1]*hop_length, class_labels[col]))
    fid.close()
```

**Comment by leanderme** [comment]
Thank you for the answer! :tada: This helps a lot!

Two things that I noticed: (1) using your code was my first intuition, too. But this limits the prediction to a maximum offset of the last predicted class (in my case) of 0.4 seconds. See these two visualizations of the prediction output for example: 
![one_sequence](https://user-images.githubusercontent.com/4583896/57160369-dda61480-6de8-11e9-88e0-5be655af3ba5.png)
![Screenshot from 2019-05-03 16-50-52](https://user-images.githubusercontent.com/4583896/57160375-e0a10500-6de8-11e9-965a-19ad73a0a2f8.png)

The prediction is fine. But at this moment, I realized that the model actually only runs for one sequence (defined as `seq_len`) of the mel-spectrogram.

(2) So I thought that I had to append each sequence into one input matrix as the input for the prediction. This was when I had a closer look on the DataGenerator class of the DCASE 2019 repository. When I run the prediction on the generated input matrix, it seems to consider all the sequences for one input file. But the predictions are different from using your described method, and are just wrong. 

Anyhow, here is the full script, using your last suggested code:

```python
import os
import numpy as np
import utils
import keras.backend as K
from sklearn.externals import joblib
import matplotlib.pyplot as plt
from keras.models import model_from_json
import parameter # dict containing config parameters
from collections import deque

# functions for file operations
# all audio files are zero-padded to match the longest one in the set
from file import *

K.set_image_data_format('channels_first')

params = parameter.get_params()

def contiguous_regions(activity_array):
  # Find the changes in the activity_array
  change_indices = np.diff(activity_array).nonzero()[0]

  # Shift change_index with one, focus on the frame after the change.
  change_indices += 1

  if activity_array[0]:
      # If the first element of activity_array is True add 0 at the beginning
      change_indices = np.r_[0, change_indices]

  if activity_array[-1]:
      # If the last element of activity_array is True, add the length of the array
      change_indices = np.r_[change_indices, activity_array.size]

  # Reshape the result into two columns
  return change_indices.reshape((-1, 2))

def test(audio_file, model_json, weights):
  nb_ch = 1
  batch_size = params['batch_size']
  seq_len = params['seq_len']
  is_mono = params['is_mono']
  __class_labels = params['__class_labels']
  evaluation_setup_folder = './evaluation_setup'

  sample_rate = sr = params['sample_rate']

  nfft = params['fft_size']
  hop_len = params['hop_size']
  nb_mel_bands = params['nb_mel_bands']

  #Loads model from Json file
  model = model_from_json(open(model_json).read())
  #Loads pre-trained weights for the model
  model.load_weights(weights)
  #Compiles the model
  model.compile(optimizer='Adam', loss='binary_crossentropy')
  model.summary()


  y, sr, silence_entry, original_len = load_audio(audio_file, mono=is_mono, fs=sr)
  mbe = melspectrogram(y, sr, nfft, nb_mel_bands, hop_len)['feat']

  scaler = joblib.load("my_scaler.save") 
  feat_test = scaler.transform(mbe)

  # split into sequencse and multichannels
  feat_test = utils.split_in_seqs(feat_test, seq_len)
  feat_test = utils.split_multi_channels(feat_test, nb_ch)

  thresh = 0.5
  hop_length = 0.02

  # Predict
  pred_file = model.predict(feat_test)
  pred_file_threshold = (pred_file > thresh).astype(int)
  pred_file_threshold = utils.reshape_3Dto2D(pred_file_threshold)

  fid = open("results.txt", 'w')
  for col in range(pred_file_threshold.shape[1]):
    # For each class, identify all onset-offset times of its instances
    region_list = contiguous_regions(pred_file_threshold[:, col])
    for region in region_list:
      fid.write('{}\t{}\t{}\t{}\n'.format(
        audio_file,
        region[0] * hop_length,
        region[1] * hop_length,
        list(__class_labels.keys())[col])
      )

  fid.close()

if __name__ == '__main__':
  # model.json contains the model defintion
  test('file.wav', 'model.json', 'model.h5')
```

**Comment by sharathadavanne** [comment]
I did not understand your problem clearly.  All I could figure out was your prediction time-boundaries are not perfect yet. I am guessing you have not tuned the model for your dataset. You might need to tune the model parameters, start with the number of layers (add or remove a layer) and the number of units per layer (I generally increase/decrease in power of 2). Next play around with the input sequence length (again power of 2). Generally, with the current training procedure of CRNN model, the time boundary detection is pretty good with most datasets I have handled.

**Comment by leanderme** [comment]
I'm sorry, you are right! I took a step backwards and trained a model on the DCASE 2017 task 3 data and can confirm that the code above works. My model isn't trained well which caused the confusion.

Thank you so much for the help and assistance.

**Comment by n33lkanth** [comment]
@leanderme @sharathadavanne  Hi,
I am also working on re-implementing the DCASE 2017 Task 2 (Monophonic SED task) but with a different model. I would like to know the process to use the Event-based evaluation matric. It will be a great help if you can guide me to implement the evaluation metric. Also, is there any implementation for DCASE 2017 Task 2 apart from the baseline method?


Best Regards
